File: backend/app/schemas/product.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List, Dict, Any
from datetime import datetime
from decimal import Decimal

class ProductBase(BaseModel):
    """Product 기본 스키마"""
    name: str = Field(..., description="상품명")
    description: str = Field(..., description="상품 설명")
    price: Decimal = Field(..., gt=0, description="상품 가격")
    category: str = Field(..., description="상품 카테고리")
    brand: Optional[str] = Field(None, description="브랜드")
    image_url: Optional[str] = Field(None, description="상품 이미지 URL")
    tags: Optional[List[str]] = Field(default=[], description="상품 태그")
    attributes: Optional[Dict[str, Any]] = Field(default={}, description="상품 속성")
    is_active: bool = Field(True, description="상품 활성화 여부")

class ProductCreate(ProductBase):
    """Product 생성 스키마"""
    pass

class ProductUpdate(BaseModel):
    """Product 업데이트 스키마"""
    name: Optional[str] = Field(None, description="상품명")
    description: Optional[str] = Field(None, description="상품 설명")
    price: Optional[Decimal] = Field(None, gt=0, description="상품 가격")
    category: Optional[str] = Field(None, description="상품 카테고리")
    brand: Optional[str] = Field(None, description="브랜드")
    image_url: Optional[str] = Field(None, description="상품 이미지 URL")
    tags: Optional[List[str]] = Field(None, description="상품 태그")
    attributes: Optional[Dict[str, Any]] = Field(None, description="상품 속성")
    is_active: Optional[bool] = Field(None, description="상품 활성화 여부")

class ProductInDB(ProductBase):
    """Product DB 스키마"""
    model_config = ConfigDict(from_attributes=True)
    
    product_no: str = Field(..., description="상품 번호 (Primary Key)")
    created_at: datetime = Field(..., description="생성일시")
    updated_at: datetime = Field(..., description="수정일시")
    rating: Optional[float] = Field(None, ge=0, le=5, description="평점")
    review_count: int = Field(0, description="리뷰 수")
    view_count: int = Field(0, description="조회수")
    sales_count: int = Field(0, description="판매량")
    status: str = Field("active", description="상품 상태")

class Product(ProductInDB):
    """Product 응답 스키마"""
    pass

class ProductList(BaseModel):
    """Product 목록 응답 스키마"""
    items: List[Product]
    total: int
    page: int
    size: int
    total_pages: int

class ProductSearch(BaseModel):
    """Product 검색 요청 스키마"""
    query: Optional[str] = Field(None, description="검색어")
    category: Optional[str] = Field(None, description="카테고리 필터")
    category_id: Optional[int] = Field(None, description="카테고리 ID 필터")
    brand: Optional[str] = Field(None, description="브랜드 필터")
    min_price: Optional[Decimal] = Field(None, ge=0, description="최소 가격")
    max_price: Optional[Decimal] = Field(None, ge=0, description="최대 가격")
    min_rating: Optional[float] = Field(None, ge=0, le=5, description="최소 평점")
    tags: Optional[List[str]] = Field(None, description="태그 필터")
    sort_by: Optional[str] = Field("created_at", description="정렬 기준")
    sort_order: Optional[str] = Field("desc", description="정렬 순서")

class ProductStats(BaseModel):
    """Product 통계 스키마"""
    total_products: int
    total_categories: int
    total_brands: int
    average_price: Decimal
    average_rating: float
    total_reviews: int
    total_sales: int

class ProductStatistics(BaseModel):
    """상품별 통계 스키마"""
    product_no: int = Field(..., description="상품 번호")
    total_reviews: int = Field(0, description="총 리뷰 수")
    average_rating: float = Field(0.0, ge=0, le=5, description="평균 평점")
    rating_distribution: str = Field("{}", description="평점 분포 (JSON)")
    last_review_date: str = Field("", description="마지막 리뷰 날짜")
    review_velocity: float = Field(0.0, description="월평균 리뷰 수")

class Category(BaseModel):
    """카테고리 스키마"""
    category_id: int = Field(..., description="카테고리 ID")
    category_name: str = Field(..., description="카테고리명")
    category_code: Optional[str] = Field(None, description="카테고리 코드")
    parent_category_id: Optional[int] = Field(None, description="상위 카테고리 ID")
    depth: int = Field(0, description="카테고리 깊이")
    created_at: Optional[datetime] = Field(None, description="생성일시")
    updated_at: Optional[datetime] = Field(None, description="수정일시")

class CategoryList(BaseModel):
    """카테고리 목록 응답 스키마"""
    items: List[Category]
    total: int 

File: frontend/src/services/product.ts
const API_BASE_URL = 'http://localhost:8000/api/v1';

export interface Product {
  id: string;
  product_no: string;
  name: string;
  price: number;
  rating?: number;
  image_url?: string;
  category: string;
  description: string;
  review_count: number;
  brand?: string;
  tags?: string[];
  created_at: string;
  updated_at: string;
}

export interface Category {
  category_id: number;
  category_name: string;
  category_code?: string;
  parent_category_id?: number;
  depth: number;
  created_at?: string;
  updated_at?: string;
}

export interface CategoryList {
  items: Category[];
  total: number;
}

export interface ProductStatistics {
  product_no: number;
  total_reviews: number;
  average_rating: number;
  rating_distribution: string;
  last_review_date: string;
  review_velocity: number;
}

export interface ProductSearchParams {
  query?: string;
  category?: string;
  category_id?: number;
  brand?: string;
  min_price?: number;
  max_price?: number;
  min_rating?: number;
  tags?: string[];
  sort_by?: string;
  sort_order?: string;
}

export interface ProductList {
  items: Product[];
  total: number;
  page: number;
  size: number;
  total_pages: number;
}

export interface ReviewSearchParams {
  query: string;
  page?: number;
  size?: number;
  hybrid_weight?: number;
}

class ProductService {
  private getAuthHeaders(): Record<string, string> {
    const token = localStorage.getItem('access_token');
    return token ? { 'Authorization': `Bearer ${token}` } : {};
  }

  async searchProducts(params: ProductSearchParams, page: number = 1, size: number = 10): Promise<ProductList> {
    const searchParams = new URLSearchParams();
    
    // 검색 파라미터 추가
    if (params.query) searchParams.append('query', params.query);
    if (params.category) searchParams.append('category', params.category);
    if (params.category_id) searchParams.append('category_id', params.category_id.toString());
    if (params.brand) searchParams.append('brand', params.brand);
    if (params.min_price) searchParams.append('min_price', params.min_price.toString());
    if (params.max_price) searchParams.append('max_price', params.max_price.toString());
    if (params.min_rating) searchParams.append('min_rating', params.min_rating.toString());
    if (params.tags) searchParams.append('tags', params.tags.join(','));
    if (params.sort_by) searchParams.append('sort_by', params.sort_by);
    if (params.sort_order) searchParams.append('sort_order', params.sort_order);
    
    // 페이지네이션 파라미터
    searchParams.append('page', page.toString());
    searchParams.append('size', size.toString());

    const response = await fetch(`${API_BASE_URL}/products/search?${searchParams.toString()}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        ...this.getAuthHeaders(),
      },
    });

    if (!response.ok) {
      throw new Error('Failed to search products');
    }

    return await response.json();
  }

  async searchProductsByReviews(query: string, page: number = 1, size: number = 10): Promise<ProductList> {
    const response = await fetch(`${API_BASE_URL}/reviews/search-products-by-reviews`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.getAuthHeaders(),
      },
      body: JSON.stringify({
        query: query,
        page: page,
        size: size,
        min_rating: 3.0,
        hybrid_weight: 0.6
      }),
    });

    if (!response.ok) {
      throw new Error('Failed to search products by reviews');
    }

    return await response.json();
  }

  async getProductById(productNo: string): Promise<Product> {
    const response = await fetch(`${API_BASE_URL}/products/${productNo}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        ...this.getAuthHeaders(),
      },
    });

    if (!response.ok) {
      throw new Error('Failed to get product');
    }

    return await response.json();
  }

  async getSimilarProducts(productNo: string, size: number = 10): Promise<ProductList> {
    const response = await fetch(`${API_BASE_URL}/products/similar/${productNo}?size=${size}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        ...this.getAuthHeaders(),
      },
    });

    if (!response.ok) {
      throw new Error('Failed to get similar products');
    }

    return await response.json();
  }

  async getProductStatistics(productIds: string[]): Promise<ProductStatistics[]> {
    const response = await fetch(`${API_BASE_URL}/products/statistics`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.getAuthHeaders(),
      },
      body: JSON.stringify({ product_ids: productIds }),
    });

    if (!response.ok) {
      throw new Error('Failed to get product statistics');
    }

    return await response.json();
  }

  async getCategories(): Promise<CategoryList> {
    const response = await fetch(`${API_BASE_URL}/products/categories`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        ...this.getAuthHeaders(),
      },
    });

    if (!response.ok) {
      throw new Error('Failed to fetch categories');
    }

    return await response.json();
  }
}

export const productService = new ProductService(); 